[Daffa Fathani] iP

zunedz opened the pull request:
## Chad Duke
> “Your mind is for having ideas, not holding them.” – David Allen (source)

### Chad Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~FAST~ *SUPER* **FAST** to use

### All you need to do is,

1. download it from [here](https://github.com/zunedz/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

### And it is **FREE!**

### Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

### How to add a task:
1. First, run the application by changing your directory to be the same as the .jar file that you have just downloaded
2. Run `java -jar ip.jar`
3. Try out some commands! example: `todo eat breakfast`, `deadline do CS2103 homework /by 2022-02-02`

### If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Duke {
    public static void main(String[] args) {
          new Duke().run();
    }
}
```

Commit:
Add Gradle support

zunedz committed:
include .class in gitignore

zunedz committed:
implemented duke basic commands, greet, echo and exit

zunedz committed:
removed indentation and finished level 2

zunedz committed:
refactor code and finished level 3

zunedz committed:
implemented event, deadline and todo

zunedz committed:
added test cases to /text-ui-test/input.txt, passed

zunedz committed:
implemented error handling, fixed previous bugs

zunedz committed:
fixed bugs

zunedz committed:
added delete command, updated test cases

zunedz committed:
fixed error message and refactor codes

zunedz committed:
implemented persistent data each time Duke start, havent updated test cases

zunedz committed:
implement date parsing (minimal)

zunedz committed:
Merge branch 'branch-Level-7'

zunedz committed:
Merge branch 'branch-Level-8'

zunedz committed:
modify previous changes to implement date parsing

zunedz committed:
include /data in .gitignore and removed unused import statement

zunedz committed:
refactor code, add separate database class

zunedz committed:
more code refactoring

zunedz committed:
refactor code into Ui, Storage, TaskList

zunedz committed:
refactor code, add Parser

zunedz committed:
refactor code, add Parser

zunedz committed:
put all classes in the respective packages

zunedz committed:
add 5 JUnit tests

zunedz committed:
created .jar file

zunedz committed:
added javadoc for > half of the public methods

zunedz committed:
tweak the code to comply the coding standard

zunedz committed:
implement feature for user to find for a keyword

zunedz committed:
Merge branch 'branch-A-JavaDoc'

zunedz committed:
merge branch-A-CodingStandard

zunedz committed:
merge branch-Level-9

zunedz committed:
change some methods' name to be more intuitive

zunedz committed:
remove /data folder

zunedz committed:
Merge branch 'add-gradle-support' of github.com:zunedz/ip into A-Gradle

zunedz committed:
remove META-INF and added gradle support

zunedz committed:
remove META-INF and added gradle support

zunedz committed:
implement GUI for the chat bot using JavaFX

zunedz committed:
Merge pull request #1 from zunedz/Level-10

implement GUI for the chat bot using JavaFX

fredtwt commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    public int numTasks() {
+        return this.index;
+    }
```
Should this method name be a verb instead? (e.g. getNumberOfTasks)

zunedz replied:
good catch! thanks for the suggestion, i'll make changes accordingly

fredtwt commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,29 @@
+    public Event(String taskName, boolean isDone, String time) {
+        super(taskName, isDone);
+        this.time = LocalDate.parse(time);
+    }
```
Good job on using the "is" prefix for the boolean variable

fredtwt commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+        return out;
+    }
+
+    public String findKeyWord(String word) {
+        String output = "Here are the matching tasks in your list:";
+        for (int i = 0; i < index; i++) {
+            if (tasks.get(i).contains(word)) {
+                output += String.format("\n%d.%s", i + 1, tasks.get(i));
+            }
+        }
+        return output;
+    }
+}
```
good job here! sticking to the proper coding standards! 👍

fredtwt commented on src/main/java/duke/util/Constants.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class Constants {
+    public static final String HORIZONTAL_LINE = "____________________________________________________________";
+
+    public static final String GREETINGS = "Hello there, I am Giga Chad Duke\nHow can I help you?";
+
+    public static final String BYE = "Bye, hope to see you again soon!";
+
+    public static final List<String> TASKS = Arrays.asList("list", "bye", "todo", "event",
+            "deadline", "mark", "unmark", "delete", "find");
+
+    public static final String DATA_DIRECTORY = "./data";
+
+    public static final String FILE_NAME = "duke.txt";
+
+    public static final String TEMP_FILE_NAME = "temp.txt";
```
good job! Adhering to the naming conventions of constants

ddx-510 replied:
Agreed with @fredtwt, `Constants` class makes the code clearer and much more readable. 👍

fredtwt reviewed (commented):
I would say that overall it was a really good job with great effort! You've managed to adhere to most of the coding standards and that's a good accomplishment! Keep up the good work! 💯

ddx-510 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+     * @param word keyword search
+     * @return keyword exists inside the string taskname
+     */
+    public boolean contains(String word) {
```
Boolean methods should be named to sound like booleans, so maybe we could modify it to `hasWord` to make it more readable :)

ddx-510 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+     * @return String representation of the current state of the task.
+     */
+    public String getStatusIcon() {
+        return isDone? "X": " ";
```
Small issue here, there should be a space after the `isDone` :)

ddx-510 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,50 @@
+                throw new InvalidArgumentException();
+            }
+
+            if (input[0].equals("mark") || input[0].equals("unmark")) {
```
Would it be better if we encapsulate `input[0]` into a variable, and use the `switch` statement here instead of a long list of `if` and `else if`?     :)

zunedz replied:
yes, I agree, thanks for the input!

ddx-510 reviewed (commented):
LGTM! 👍 
Overall your codes are readable and clear. The high quality of code has impressed me a lot. 
Just some small issues to be fixed.

zunedz committed:
Parser's parseCommand method: uses nested if-else

Nested if-else statement does not follow good coding
quality (arrowhead style).

Let's change the if-else statement with switch
statement instead

Using switch statement is preferable as the code
will look way cleaner compared to nested if-else
statement (good code quality)

zunedz committed:
Parser's parse method: interact with tasklist

Parser class should only do its job, that is parsing
the input without having any access to tasks.

Let's decouple this part of the Parser class by
creating a new abstract class called Command.

By creating a new abstract class with the only job
is to interact with taskList, now Parser's class
only have to deal with parsing inputs, adhering
single responsibility principle.

zunedz committed:
remove unnecessary function argument

zunedz committed:
change build.gradle mainClassName

zunedz committed:
Use assertions to document important assumptions

zunedz committed:
Merge pull request #2 from zunedz/A-Assertions

Use assertions to document important assumptions

zunedz committed:
Improve code quality, remove arrow-head shaped methods, change function names

zunedz committed:
Merge pull request #3 from zunedz/A-CodeQuality

Improve code quality, remove arrow-head shaped methods, change functi…

zunedz committed:
implement undo command for add and delete, todo: mark and unmark

zunedz committed:
Merge pull request #4 from zunedz/C-Undo

implement undo command for add and delete, todo: mark and unmark

zunedz committed:
Code refactor, improve code quality and implement undo for mark and unmark commands

zunedz committed:
polish the GUI

zunedz committed:
change dialog box label

zunedz committed:
fix gui bug and set goodbye timer

zunedz committed:
fix assertions and exceptions

zunedz committed:
add JUnit tests with stubs

zunedz committed:
add credit - reuse policy

zunedz committed:
Add Ui.png

zunedz committed:
Update user guide

zunedz committed:
Create _config.yml

zunedz committed:
add checkstyle and edit some code

zunedz committed:
Merge branch 'master' of github.com:zunedz/ip

zunedz committed:
change .jar file name

zunedz committed:
Update README.md

zunedz committed:
Update README.md

zunedz committed:
add help command

zunedz committed:
fix indentation level

zunedz committed:
fix typo

zunedz committed:
Update README.md

zunedz committed:
fix checkstyle

zunedz committed:
Merge branch 'master' of github.com:zunedz/ip
